pkg/destroy/azure/azure.go

```diff
@@ -226,9 +247,82 @@ func (o *ClusterUninstaller) Run() (*types.ClusterQuota, error) {
 		o.Logger.Debug(err)
 	}
 
+	deadline, _ = waitCtx.Deadline()
+	diff = time.Until(deadline)
+	if diff > 0 {
+		waitCtx, cancel = context.WithTimeout(context.Background(), diff)
+	}
+
+	if err := removeSharedTags(waitCtx, o.vnetClient, o.resourceGraphClient, o.InfraID, o.SubscriptionID); err != nil {
+		errs = append(errs, fmt.Errorf("failed to remove shared tags: %w", err))
+		o.Logger.Debug(err)
+	}
+
 	return nil, utilerrors.NewAggregate(errs)
 }
 
+func removeSharedTags(
+	ctx context.Context,
+	vnetClient *armnetwork.VirtualNetworksClient,
+	graphClient *armresourcegraph.Client,
+	infraID, subscriptionID string,
+) error {
+	tagKey := fmt.Sprintf("kubernetes.io_cluster.%s", infraID)
+	query := fmt.Sprintf(
+		"resources | where type == 'microsoft.network/virtualnetworks' and tags.['%s'] == 'shared' | project name, resourceGroup, tags",
+		tagKey,
+	)
+	results, err := graphClient.Resources(ctx,
+		armresourcegraph.QueryRequest{
+			Query: &query,
+			Subscriptions: []*string{
+				&subscriptionID,
+			},
+		},
+		nil,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to query vnet: %w", err)
+	}
+
+	// marshal and unmarshal results.Data to avoid casting every type along the way.
+	// the sdk forcing us to do this is a shortcoming on its part which has been
+	// reported but so far no solution have been provided.
+	// see https://github.com/Azure/azure-sdk-for-go/issues/21004 for details.
+	b, err := json.Marshal(results.Data)
+	if err != nil {
+		return fmt.Errorf("failed to json marshall results data: %s", results.Data)
+	}
+	vnetsResult := []struct {
+		Name          string
+		ResourceGroup string
+		Tags          map[string]string
+	}{}
+	if err := json.Unmarshal(b, &vnetsResult); err != nil {
+		return fmt.Errorf("failed to json marshall results data: %s", results.Data)
+	}
+
+	for _, vnet := range vnetsResult {
+		delete(vnet.Tags, tagKey)
+
+		newTags := armnetwork.TagsObject{
+			Tags: map[string]*string{},
+		}
+
+		for k, v := range vnet.Tags {
+			newTags.Tags[k] = to.StringPtr(v)
+		}
+
+		if _, err := vnetClient.UpdateTags(
+			ctx, vnet.ResourceGroup, vnet.Name, newTags, nil,
+		); err != nil {
+			return fmt.Errorf("failed to update virtual network tags: %w", err)
+		}
+	}
```

Comment:
Can we use a more general procedure to remove the shared tags, maybe https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armresources#ClientFactory.NewTagsClient ?

Reply:
https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armresources#TagsClient.UpdateAtScope

Reply:
I think we need to both, remove it from the vnet then delete it with https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armresources#TagsClient.Delete.
The delete may fail if there's any other resource using the tag.

Reply:
We don't want to delete the tag, we want to update the resource's tags by removing the key-pair. That's way I suggested `TagsClient.UpdateAtScope`.

Reply:
done. PTAL

Reply:
Sorry, I completely missed your comment about `UpdateAtScope`. having a look.